[Docs]: Highlighting code snippets in .rst docs

0xisk opened the pull request:
### What did I change? 

- Enhancing all the .rst doc files by adding highlighting for the code snippets, including the following langs:
1. solidity
2. assembly
3. bash
4. javascript

Thanks so much!

cameel commented on docs/contributing.rst:
```diff
@@ -237,7 +237,7 @@ Automatically updating the test above changes it to
 and re-run the test. It now passes again:
 
-::
+.. code-block:: bash
```
```suggestion
.. code-block:: text
```

cameel commented on docs/contributing.rst:
```diff
@@ -273,7 +273,7 @@ Next, build Solidity (or just the ``solfuzzer`` binary) with AFL as your compile
 At this stage you should be able to see a message similar to the following:
 
-::
+.. code-block:: bash
```
```suggestion
.. code-block:: text
```

cameel commented on docs/contributing.rst:
```diff
@@ -293,7 +293,7 @@ If the instrumentation messages did not appear, try switching the cmake flags po
 Otherwise, upon execution the fuzzer halts with an error saying binary is not instrumented:
 
-::
+.. code-block:: bash
```
```suggestion
.. code-block:: text
```

cameel commented on docs/contributing.rst:
```diff
@@ -334,7 +334,7 @@ that result in similar behaviour of the binary.
 Now run the fuzzer (the ``-m`` extends the size of memory to 60 MB):
 
-::
+.. code-block:: bash
```
```suggestion
.. code-block:: text
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -87,7 +87,7 @@ Certain optimizer steps symbolically track the storage and memory locations. For
 information is used to compute Keccak-256 hashes that can be evaluated during compile time. Consider
 the sequence:
 
-::
+.. code-block:: assembly
 
   PUSH 32
   PUSH 0
```
Not sure if it's a good idea. It's not x86 assembly.

0xisk replied:
I agree with you, but the intention of using it here is to have the following differences. So I think it's more readable. 
![pic-selected-210604-0614-34](https://user-images.githubusercontent.com/36774944/120744905-52823600-c4fc-11eb-8971-6e3fdbb48c4e.png) ![pic-selected-210604-0614-12](https://user-images.githubusercontent.com/36774944/120744908-531acc80-c4fc-11eb-97e6-ca3d14a8a7c2.png)

cameel replied:
It's not that bad in this particular example, though still the color difference between `PUSH` and `DUP2` might make users think that they're different in some important way. In bigger samples it's even worse. Personally, I'd not highlight this one.

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -99,7 +99,7 @@ the sequence:
 or the equivalent Yul
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -114,7 +114,7 @@ the instruction doesn't write to a certain location.
 For example,
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

Most (though not all) snippets in this file contiain Yul code rather than Solidity.

0xisk replied:
It's my lack of experience in Yul language, it's pretty much similar to the Assembly code in Solidity. So that's why I regarded them as Solidity.

I will check this script now.

0xisk replied:
@cameel @chriseth  My point of view for choosing `.. code-block:: solidity` rather than using `.. code-block:: yul` is to have the code highlighted in Github, since using `yul` won't highlight the code.

Below is the difference between using both: 
- `.. code-block:: solidity`
![pic-selected-210608-0736-26](https://user-images.githubusercontent.com/36774944/121129310-0d783f80-c82d-11eb-8b9d-1acdb081710f.png)

- `.. code-block:: yul`
![pic-selected-210608-0736-47](https://user-images.githubusercontent.com/36774944/121129335-1537e400-c82d-11eb-8d0d-958e6978605f.png)

cameel replied:
I understand but it will also switch to this inferior highlighting in the official docs (https://docs.soliditylang.org) where we do have proper Yul highlighting already and I think that takes priority.

0xisk replied:
Alrigth great I will switch it to `yul`

cameel commented on docs/types/conversion.rst:
```diff
@@ -59,28 +59,36 @@ At the end of this code snippet, ``x`` will have the value ``0xfffff..fd`` (64 h
-after the conversion (if the index is still in range)::
+after the conversion (if the index is still in range):
+
+.. code-block:: solidity
```
I'm a bit suprised that `::` at the end of a sentence was even valid syntax.

0xisk replied:
Yes I agree with you

cameel reviewed (commented):
Thanks for the PR! Here's some feedback from me:
- Some of these snippets are mislabeled. For example some `text` shippets are marked as `bash` and some `solidity` ones are really `yul`.
- Should we use `bash` or `shell` for shell snippets?
- I'm not sure if we actually want to replace all the `::` bits with `code-block::`. I personally prefer the `code-block::` but `::` is already pretty prevalent so we should first discuss this change.
- Please squash your commits. Small commits are nice but simply putting every file in a separate commit does not really help reviewing. One way to split it that would actually be helpful would be if different highlighting languages were in different commits. But a single commit would be fine too since all of these modifications are closely related.

Also, @axic mentioned at some point that he already had some stashed work cleaning this up so this might conflict with his changes.

chriseth commented:
@cameel can you check that the script that extracts code snippets from the documentation still works with these changes? I assume it does, but please double-check.

cameel commented:
Sure. I'll check it when I get back to finish #11420.

iskanderandrews commented:
> Thanks for the PR! Here's some feedback from me:

Thanks so much for your feedback. 
 
> * Some of these snippets are mislabeled. For example some `text` shippets are marked as `bash` and some `solidity` ones are really `yul`.

Regarding Yul and Solididyt conflicting, it was just because of my lack of experience with Yul. I will go over the changes again to make sure of that. 

> * Should we use `bash` or `shell` for shell snippets?

For my I based my selection to the most used one, so before my changes `.. code-block:: bash` was used more in all the docs: 

1. 4 times in [docs/smtchecker.rst](https://github.com/ethereum/solidity/blob/develop/docs/smtchecker.rst)
2. 2 times in [docs/layout-of-source-files.rst](https://github.com/ethereum/solidity/blob/develop/docs/layout-of-source-files.rst)
3. 23 times in [docs/installing-solidity.rst](https://github.com/ethereum/solidity/blob/develop/docs/installing-solidity.rst)

While `.. code-block:: bash` was used just 1 time in [docs/using-the-compiler.rst](https://github.com/ethereum/solidity/blob/develop/docs/using-the-compiler.rst)

Also, based on the following [answer](https://stackoverflow.com/a/5725402/9058556) from StackOverflow on the `Difference between sh and bash`: 

> sh = bash?
For a long time, /bin/sh used to point to /bin/bash on most GNU/Linux systems. As a result, it had almost become safe to ignore the difference between the two. But that started to change recently.

So I think we could follow with `.. code-block:: bash` since it's the most used one already. 

> * I'm not sure if we actually want to replace all the `::` bits with `code-block::`. I personally prefer the `code-block::` but `::` is already pretty prevalent so we should first discuss this change.

Based on the **rst** [documentation](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/index.html) in this section:

> To show example code, use the reStructured Text `code-block` directive: [Show Example Code](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/code_blocks.html#show-example-code)

So in order to highlight an example snippet of code, we need to use  `code-block`

> * Please squash your commits. Small commits are nice but simply putting every file in a separate commit does not really help reviewing. One way to split it that would actually be helpful would be if different highlighting languages were in different commits. But a single commit would be fine too since all of these modifications are closely related.

Yes, I agree with you, I'm working on squashing commits now.

iskanderandrews commented:
> * Please squash your commits. Small commits are nice but simply putting every file in a separate commit does not really help reviewing. One way to split it that would actually be helpful would be if different highlighting languages were in different commits. But a single commit would be fine too since all of these modifications are closely related.

Commits squashing is done.

chriseth commented on docs/analysing-compilation-output.rst:
```diff
@@ -23,7 +23,7 @@ Consider the following contract (named, say ``contract.sol``):
 The following would be the output of ``solc --asm contract.sol``
 
-.. code-block:: none
+.. code-block:: assembly
```
I think this should rather be "plain text", I don't think "assembly" highlights properly.

chriseth replied:
Ah, just saw the comments above. I'll have a look at the preview to check if it looks nice.

iskanderandrews replied:
My point of view for choosing `.. code-block:: solidity` rather than using `.. code-block:: yul` is to have the code highlited in Github, since using `yul` won't highlight the code.

Below is the difference between using both: 
- `.. code-block:: solidity`
![pic-selected-210608-0736-26](https://user-images.githubusercontent.com/36774944/121129310-0d783f80-c82d-11eb-8b9d-1acdb081710f.png)

- `.. code-block:: yul`
![pic-selected-210608-0736-47](https://user-images.githubusercontent.com/36774944/121129335-1537e400-c82d-11eb-8d0d-958e6978605f.png)

cameel replied:
I'm not sure this is good highlighting. Colors are a bit random. Might mislead some users into thinking that some things have special meaning when they do not. My vote would be for plain text for this.

0xisk replied:
I agree with you @cameel that might mislead some users, I will switch to `.. code-block:: none`

chriseth commented:
`conf.py` mentions `highlight_language = 'Solidity'` - can we somehow rely on this default so that we do not have to specify the language for all the solidity code examples?

iskanderandrews commented:
> `conf.py` mentions `highlight_language = 'Solidity'` - can we somehow rely on this default so that we do not have to specify the language for all the solidity code examples?

The intention to have the code highlighted in GitHub, to be more readable.

cameel commented:
@iskanderandrews Note that the docs are primarily meant to be viewed after being rendered by Sphinx (e.g. on https://docs.soliditylang.org), for which we do have Yul highlighting via a pygments plugin. Good highlighting on github is a good thing if possible but if there's a conflict between the too, I'd rather choose good highlighting in Sphinx over github.

But I think it's a good argument for not relying on the default language in `conf.py`. If github highlights it as solidity only if we specify the language explicitly then I think it's better to do so.

chriseth commented:
What do you mean by "highlighted in github"? Can you link to an example page on github?

iskanderandrews commented:
> What do you mean by "highlighted in github"? Can you link to an example page on github?
Sure, so this is an example `docs/common-patterns.rst` from both repos: 

- Main ethereum/solidity: 
https://github.com/ethereum/solidity/blob/develop/docs/common-patterns.rst

- Forked version iskanderandrews/solidity
https://github.com/iskanderandrews/solidity/blob/feature/docs-highlight-solc/docs/common-patterns.rst

So that's what I mean by "highlighted in github" thus making it more readable.

cameel commented:
@chriseth 
> @cameel can you check that the script that extracts code snippets from the documentation still works with these changes? I assume it does, but please double-check.

It does. In #11565 I now have tests that show that it works both with `::` and `.. code-block:: solidity`. It does not really have any of them hard-coded. It finds all indented blocks in the docs and uses a heuristic to determine if they're Solidity code (i.e. checks for the SPDX comment, version pragma of a contract/library/interface declaration).

cameel commented on docs/assembly.rst:
```diff
@@ -216,7 +216,9 @@ starting from where this pointer points at and update it.
-Here is an assembly snippet you can use for allocating memory that follows the process outlined above::
+Here is an assembly snippet you can use for allocating memory that follows the process outlined above
+
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

alice-dfi commented:
Wow what are people doing blocking codes can someone explain ? It better
not be any of my contracts , I see your blocking and deleting files and
contracts like left and right ????

On Mon, Jun 21, 2021 at 3:10 PM Kamil Śliwak ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In docs/analysing-compilation-output.rst
> <https://github.com/ethereum/solidity/pull/11477#discussion_r655637670>:
>
> > @@ -23,7 +23,7 @@ Consider the following contract (named, say ``contract.sol``):
>
>  The following would be the output of ``solc --asm contract.sol``
>
> -.. code-block:: none
> +.. code-block:: assembly
>
> I'm not sure this is good highlighting. Colors are a bit random. Might
> mislead some users into thinking that some things have special meaning when
> they do not. My vote would be for plain text for this.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/11477#discussion_r655637670>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUBR3DPZMW7Z46AMVRCCZL3TT6FCNANCNFSM457G2W2Q>
> .
>

cameel commented on docs/contracts/errors.rst:
```diff
@@ -79,4 +79,4 @@ of the built-in type ``Panic(uint256)``.
     can "forge" revert data that looks like it could have come from the
-    contract that called it.
\ No newline at end of file
+    contract that called it.
```
Is there actually a difference here? If it was a newline github would show that but it does not so I wonder what's going on here.

0xisk replied:
Yeah it's weird I didn't catch out what's changed here. I think in this script I just highlighted Solidity.

cameel commented on docs/contributing.rst:
```diff
@@ -200,7 +200,7 @@ editing of failing contracts using your preferred text editor. Let's try to brea
 Running ``./build/test/isoltest`` again results in a test failure:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: text
```

cameel commented on docs/contributing.rst:
```diff
@@ -334,7 +334,7 @@ that result in similar behaviour of the binary.
 Now run the fuzzer (the ``-m`` extends the size of memory to 60 MB):
 
-::
+.. code-block:: text
```
```suggestion
.. code-block:: bash
```

cameel commented on docs/examples/micropayment.rst:
```diff
@@ -90,7 +90,7 @@ library provides a function called ``soliditySHA3`` that mimics the behaviour of
 Here is a JavaScript function that creates the proper signature for the ``ReceiverPays`` example:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: javascript
```

alice-dfi commented:
Who wants to make a million dollars right now start making some of my files on GitHub and block some scammers ... contact me 2018568555 simple as that .

cameel commented on docs/examples/micropayment.rst:
```diff
@@ -271,7 +271,7 @@ to prevent a message intended for one payment channel from being used for a diff
 Here is the modified JavaScript code to cryptographically sign a message from the previous section:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: javascript
```

cameel commented on docs/examples/micropayment.rst:
```diff
@@ -454,7 +454,7 @@ We'll use the `ethereumjs-util <https://github.com/ethereumjs/ethereumjs-util>`_
 and we use JavaScript. The following code borrows the ``constructMessage`` function from the signing **JavaScript code** above:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: javascript
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -460,7 +460,7 @@ so that the following expressions still only need to reference SSA variables.
 An example transformation is the following:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -476,7 +476,7 @@ An example transformation is the following:
 as follows:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -526,7 +526,7 @@ as arguments.
 The above would be transformed into
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -563,7 +563,7 @@ reassigned variables are replaced by the newly declared variables.
 Example:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -573,7 +573,7 @@ Example:
 is transformed to
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -618,7 +618,7 @@ RedundantAssignEliminator
 these might be unnecessary in many cases, like the following example:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -629,7 +629,7 @@ these might be unnecessary in many cases, like the following example:
 The SSA transform converts this snippet to the following:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -645,7 +645,7 @@ The Redundant Assign Eliminator removes all the three assignments to ``a``, beca
 snippet into strict SSA form:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -974,7 +974,7 @@ BlockFlattener
 inner block at the appropriate place in the outer block:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -986,7 +986,7 @@ inner block at the appropriate place in the outer block:
 is transformed to
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1026,7 +1026,7 @@ If a function, say, ``function f(a, b) { sstore (a, b) }``, is called with liter
 function ``f_1`` that takes only one argument, i.e.,
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1051,7 +1051,7 @@ This step removes unused parameters in a function.
 remove the parameter and create a new "linking" function as follows:
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1184,14 +1184,14 @@ fresh variable declarations.
 The SSA transform rewrites
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1184,14 +1184,14 @@ fresh variable declarations.
 to
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1202,7 +1202,7 @@ whenever ``a`` was referenced. The SSA transform changes statements
 snippet is turned into
 
-::
+.. code-block:: solidity
```
```suggestion
.. code-block:: yul
```

cameel commented on docs/ir/ir-breaking-changes.rst:
```diff
@@ -14,7 +14,8 @@ hiding new and different behavior in existing code.
 Consequently, if the padding space within a struct is used to store data (e.g. in the context of a contract upgrade), you have to be aware that ``delete`` will now also clear the added member (while it wouldn't have been cleared in the past).
 
-::
+.. code-bl::ock:: solidity
```
```suggestion
.. code-block:: solidity
```

cameel reviewed (commented):
@iskanderandrews Sorry, for not getting back to this earlier. I finally had time to finish my related PR (#11420) so now I'm also back to reviewing this.

> So I think we could follow with .. code-block:: bash since it's the most used one already.

Fair enough. Let's keep `bash`.

Apart from that there are 3 remaining issues:
1. Some blocks are still mislabeled (mostly `solidity` instead of `yul` but also `solidity` where it should be `javascript` and a few others). I pointed them all out in comments.
2. Some minor mistakes like `.. cod::e-block::` in one place.
3. I still think asm blocks should be marked as text. I mean, either way it's not a big deal but the generic asm highlighting is not very good for them in my opinion. Maybe we need another opinion here.

Once that's done please rebase on `develop` and I think we can merge it :)

Commit:
docs/
Enhancing all the .rst doc files by adding highlighting for the code snippets, including the following langs:
1. Solidity

2. bash

3. javascript

4. assembly

iskanderandrews commented:
> @iskanderandrews Sorry, for not getting back to this earlier. I finally had time to finish my related PR (#11420) so now I'm also back to reviewing this.
> 
> > So I think we could follow with .. code-block:: bash since it's the most used one already.
> 
> Fair enough. Let's keep `bash`.
> 
> Apart from that there are 3 remaining issues:
> 
> 1. Some blocks are still mislabeled (mostly `solidity` instead of `yul` but also `solidity` where it should be `javascript` and a few others). I pointed them all out in comments.
> 2. Some minor mistakes like `.. cod::e-block::` in one place.
> 3. I still think asm blocks should be marked as text. I mean, either way it's not a big deal but the generic asm highlighting is not very good for them in my opinion. Maybe we need another opinion here.
> 
> Once that's done please rebase on `develop` and I think we can merge it :)

Alright so I think all is resolved, please review again, and mention me for any other changes or edits. 
Thank you so much!

cameel reviewed (approved):
Thanks! It looks good now so I'm marking as approved.

I see that CI is not running on this PR, probably because it's an external one. @chriseth Do these need to be triggered manually now?

chriseth commented:
It should run, not sure why it does not - or maybe it just does not report its status?
